modules/optflow/src/rlof/rlof_invokerbase.hpp

```diff
@@ -0,0 +1,190 @@
+﻿// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+#ifndef _RLOF_INVOKERBASE_HPP_
+#define _RLOF_INVOKERBASE_HPP_
+
+
+#if CV_SSE4_1
+#define RLOF_SSE
+#endif
+
+//#define DEBUG_INVOKER
+
+#ifndef CV_DESCALE
+#define CV_DESCALE(x, n)     (((x) + (1 << ((n)-1))) >> (n))
+#endif
+
+#define FLT_RESCALE 1
+
+
+#include "rlof_localflow.h"
+#include <unordered_map>
+#include "opencv2/core/hal/intrin.hpp"
+using namespace std;
+using namespace cv;
+
+
+namespace cv {
+namespace optflow {
+
+typedef short deriv_type;
+#ifdef RLOF_SSE
+static inline void get4BitMask(const int & width, __m128i & mask)
+{
+    int noBits = width - static_cast<int>(floor(width / 4.f) * 4.f);
+    unsigned int val[4];
+    for (int n = 0; n < 4; n++)
+    {
+        val[n] = (noBits > n) ? (std::numeric_limits<unsigned int>::max()) : 0;
+    }
+    mask = _mm_set_epi32(val[3], val[2], val[1], val[0]);
+
+}
+static inline void getWBitMask(const int & width, __m128i & t0, __m128i & t1, __m128i & t2)
+{
+    int noBits = width - static_cast<int>(floor(width / 8.f) * 8.f);
+    unsigned short val[8];
+    for (int n = 0; n < 8; n++)
+    {
+        val[n] = (noBits > n) ? (0xffff) : 0;
+    }
+    t1 = _mm_set_epi16(val[7], val[7], val[6], val[6], val[5], val[5], val[4], val[4]);
+    t0 = _mm_set_epi16(val[3], val[3], val[2], val[2], val[1], val[1], val[0], val[0]);
+    t2 = _mm_set_epi16(val[7], val[6], val[5], val[4], val[3], val[2], val[1], val[0]);
+}
+#endif
+typedef uchar tMaskType;
+#define tCVMaskType CV_8UC1
+#define MaskSet 0xffffffff
+
+static inline bool notSameColor(const cv::Point3_<uchar> & ref, int r, int c, const cv::Mat & img, int winSize, int threshold)
+{
+    if (r >= img.rows + winSize || c >= img.cols + winSize || r < -winSize || c < -winSize) return true;
+    int step = static_cast<int>(img.step1());
+    const cv::Point3_<uchar> * tval = img.ptr<cv::Point3_<uchar>>();
+    tval += c;
+    tval += r * step / 3;
```

Comment:
One more manual pointer arithmetic:

```.diff
-int step = static_cast<int>(img.step1());
-const cv::Point3_<uchar> * tval = img.ptr<cv::Point3_<uchar>>();
-tval += c;
-tval += r * step / 3;
+CV_DbgAssert(img.type() == CV_8UC3);
+const cv::Point3_<uchar> * tval = img.ptr<cv::Point3_<uchar>>(r, c);
```

"if" check above should be updated too.

Reply:
I removed this function because it is not used anymore. Instead computation is done by the HorizontalCrossSegmentation class